### PR TITLE
feat(rwx): share manager locality

### PR DESCRIPTION
**Pull request by derekbit**
#### Which issue(s) this PR fixes:

Issue Longhorn/longhorn#7872

#### What this PR does / why we need it:

- Share manager pod respects the newly introduced `storageClass.Parameters["shareManagerNodeSelector"]`
- Share manager pod respects the `storageClass.allowedTopologies` which are internally converted to the affinity of a share manager pod
- Share manager pod respects the newly introduced `storageClass.Parameters["shareManagerTolerations"]`

### CODE REVIEW

File: controller/share_manager_controller.go [hunk 1]
```diff
@@ -798,6 +860,40 @@ func (c *ShareManagerController) createShareManagerPod(sm *longhorn.ShareManager
...
+			affinity = c.getAffinityFromStorageClass(sc)
+
+			// Find the node selector from the storage class and merge it with the system managed components node selector
+			if nodeSelector == nil {
+				nodeSelector = map[string]string{}
+			}
+			nodeSelectorFromStorageClass := c.getShareManagerNodeSelectorFromStorageClass(sc)
+			for k, v := range nodeSelectorFromStorageClass {
+				nodeSelector[k] = v
+			}
```
Comment (shuo-wu) [inline controller/share_manager_controller.go:886]: This merge means that the final scheduling should satisfy both `node selector from SC` and `system managed components node selector`, right?

**Review by james-munson — COMMENTED** [review]
~~No other questions than Shuo's.~~
Well, maybe one.  Does any of these settings merit a global Settings value?

File: controller/share_manager_controller.go (hunk 1 above)
Comment (derekbit) [reply controller/share_manager_controller.go:879]: You are right!

File: controller/share_manager_controller.go [hunk 2]
```diff
@@ -743,6 +744,36 @@ func (c *ShareManagerController) syncShareManagerPod(sm *longhorn.ShareManager)
 	return nil
 }
 
+func (c *ShareManagerController) getAffinityFromStorageClass(sc *storagev1.StorageClass) *corev1.Affinity {
+	var matchLabelExpressions []corev1.NodeSelectorRequirement
+
+	for _, topology := range sc.AllowedTopologies {
```
[the remark this answers is not in the document]
Comment (derekbit) [reply controller/share_manager_controller.go:750]: Just found longhorn doesn't handle `CreateVolumeRequest.AccessibilityRequirements (TopologyRequirement)` in CSI controller server `CreateVolume()`.
Some CSI drivers such as ceph-csi and minio-csi handle the topologies for RWO volumes. We probably can improve this part. I can create a ticket for it. WDYT?

File: controller/share_manager_controller.go [hunk 3]
```diff
@@ -798,6 +860,33 @@ func (c *ShareManagerController) createShareManagerPod(sm *longhorn.ShareManager
 		return nil, err
 	}
 
+	var affinity *corev1.Affinity
+
+	if pv.Spec.StorageClassName != "" {
+		sc, err := c.ds.GetStorageClass(pv.Spec.StorageClassName)
+		if err != nil {
+			c.logger.WithError(err).Warnf("Failed to get storage class %v", pv.Spec.StorageClassName)
```
Comment (shuo-wu) [inline controller/share_manager_controller.go:868]: NIT: "..., will continue the share manager pod creation"

**Review by shuo-wu — DISMISSED** [review]
LGTM

File: controller/share_manager_controller.go (hunk 3 above)
Comment (derekbit) [reply controller/share_manager_controller.go]: Updated. Thank you.

**Comment by derekbit** [comment]
cc @shuo-wu @PhanLe1010 @james-munson @innobead for review and approval. Thank you.

**Review by james-munson — APPROVED** [review]
Looks good.  There will need to be documentation for how to use it effectively.

**Pull request merged**